src/anm/animation/element.js

```diff
@@ -744,101 +744,141 @@ Element.prototype._checkSwitcher = function(next) {
-                pt = bounds_pts[i];
-                if (pt.x < minX) minX = pt.x;
-                if (pt.y < minY) minY = pt.y;
-                if (pt.x > maxX) maxX = pt.x;
-                if (pt.y > maxY) maxY = pt.y;
-            }
 
-            var ratio  = engine.PX_RATIO,
-                x = minX, y = minY,
-                width  = Math.round(maxX - minX),
-                height = Math.round(maxY - minY);
+            ctx.restore();
 
-            var last_cvs_size = this._maskCvsSize || engine.getCanvasSize(mcvs);
+        } else if (this.isMask) { // this element is a mask
 
-            if ((last_cvs_size[0] < width) ||
-                (last_cvs_size[1] < height)) {
-                // mcvs/bcvs both always have the same size, so we save/check only one of them
-                this._maskCvsSize = engine.setCanvasSize(mcvs, width, height);
-                engine.setCanvasSize(bcvs, width, height);
-            } else {
-                this._maskCvsSize = last_cvs_size;
-            }
+            this._renderAsMask();
 
-            var scale = ratio;  // multiple by global scale when it's known
+        } else if (this.$mask) { // this element is masked
 
-            bctx.clearRect(0, 0, width*scale, height*scale);
-            mctx.clearRect(0, 0, width*scale, height*scale);
+            ctx.save();
+            this._renderAsMasked(ctx);
+            ctx.restore();
 
-            bctx.save();
-            mctx.save();
+        }
 
-            bctx.setTransform(scale, 0, 0, scale, -x*scale, -y*scale);
-            mctx.setTransform(scale, 0, 0, scale, -x*scale, -y*scale);
+    }
 
-            this.transform(bctx);
-            this.painters(bctx);
-            this.each(function(child) {
-                child.render(bctx, dt);
-            });
+    this.__postRender();
+    this.rendering = false;
+    return this;
+};
 
-            // FIXME: this should be performed one time for all masked elements!
-            mask.transform(mctx);
-            mask.painters(mctx);
-            mask.each(function(child) {
-                child.render(mctx);
-            });
+Element.prototype._renderAsMask = function() {
+    this.ensureHasMaskCanvas();
 
-            bctx.globalCompositeOperation = 'destination-in';
-            bctx.setTransform(1, 0, 0, 1, 0, 0);
-            bctx.drawImage(mcvs, 0, 0);
+    var mcvs = this.__maskCvs,
+        mctx = this.__maskCtx,
+        bcvs = this.__backCvs;
 
-            ctx.drawImage(bcvs,
-                0, 0, Math.floor(width * scale), Math.floor(height * scale),
-                x, y, width, height);
+    // FIXME: test if bounds are not empty
+    var bounds_pts = this.bounds().toPoints();
 
-            mctx.restore();
-            bctx.restore();
-        }
-        ctx.restore();
+    var minX = Number.MAX_VALUE, minY = Number.MAX_VALUE,
+        maxX = Number.MIN_VALUE, maxY = Number.MIN_VALUE;
+
+    var pt;
+    for (var i = 0, il = bounds_pts.length; i < il; i++) {
```

Comment:
if masking element is rotated, mask bounds should fit it completely (like "bounding rectangle"). It could be an artefact now, but I am not sure yet. See [`.bounds`](https://github.com/Animatron/player/blob/refactor-masks/src/anm/animation/element.js#L2002) and [`.adaptBounds`](https://github.com/Animatron/player/blob/refactor-masks/src/anm/animation/element.js#L2106).